File: simulator/backend/db/repositories/engagement-repository.ts
/**
 * Engagement Repository
 * Data access layer for engagement metrics operations
 */

import { sql } from '../connection.ts';

export interface EngagementData {
  id: string;
  user_id: string;
  simulation_id?: string;
  event_type: 'login' | 'simulation_start' | 'simulation_complete' | 'page_view' | 'interaction';
  event_data?: any;
  timestamp: Date;
  session_id?: string;
}

export interface CreateEngagementRequest {
  user_id: string;
  simulation_id?: string;
  event_type: 'login' | 'simulation_start' | 'simulation_complete' | 'page_view' | 'interaction';
  event_data?: any;
  session_id?: string;
}

export interface EngagementStats {
  totalEvents: number;
  eventsByType: Record<string, number>;
  uniqueSessions: number;
  avgEventsPerSession: number;
}

export class EngagementRepository {
  /**
   * Find engagement by ID
   */
  async findById(id: string): Promise<EngagementData | null> {
    try {
      const result = await sql`
        SELECT id, user_id, simulation_id, event_type, event_data, timestamp, session_id
        FROM engagement_events
        WHERE id = ${id}
      `;
      return result[0] || null;
    } catch (error) {
      console.error('Error finding engagement by ID:', error);
      throw error;
    }
  }

  /**
   * Get engagement by user ID
   */
  async findByUserId(userId: string, limit?: number): Promise<EngagementData[]> {
    try {
      const query = limit
        ? sql`SELECT id, user_id, simulation_id, event_type, event_data, timestamp, session_id
             FROM engagement_events
             WHERE user_id = ${userId}
             ORDER BY timestamp DESC
             LIMIT ${limit}`
        : sql`SELECT id, user_id, simulation_id, event_type, event_data, timestamp, session_id
             FROM engagement_events
             WHERE user_id = ${userId}
             ORDER BY timestamp DESC`;

      return await query;
    } catch (error) {
      console.error('Error fetching engagement by user:', error);
      throw error;
    }
  }

  /**
   * Get engagement by simulation ID
   */
  async findBySimulationId(simulationId: string): Promise<EngagementData[]> {
    try {
      const result = await sql`
        SELECT id, user_id, simulation_id, event_type, event_data, timestamp, session_id
        FROM engagement_events
        WHERE simulation_id = ${simulationId}
        ORDER BY timestamp ASC
      `;
      return result;
    } catch (error) {
      console.error('Error fetching engagement by simulation:', error);
      throw error;
    }
  }

  /**
   * Get engagement by session ID
   */
  async findBySessionId(sessionId: string): Promise<EngagementData[]> {
    try {
      const result = await sql`
        SELECT id, user_id, simulation_id, event_type, event_data, timestamp, session_id
        FROM engagement_events
        WHERE session_id = ${sessionId}
        ORDER BY timestamp ASC
      `;
      return result;
    } catch (error) {
      console.error('Error fetching engagement by session:', error);
      throw error;
    }
  }

  /**
   * Get engagement by event type
   */
  async findByEventType(
    eventType: 'login' | 'simulation_start' | 'simulation_complete' | 'page_view' | 'interaction',
    limit?: number
  ): Promise<EngagementData[]> {
    try {
      const query = limit
        ? sql`SELECT id, user_id, simulation_id, event_type, event_data, timestamp, session_id
             FROM engagement_events
             WHERE event_type = ${eventType}
             ORDER BY timestamp DESC
             LIMIT ${limit}`
        : sql`SELECT id, user_id, simulation_id, event_type, event_data, timestamp, session_id
             FROM engagement_events
             WHERE event_type = ${eventType}
             ORDER BY timestamp DESC`;

      return await query;
    } catch (error) {
      console.error('Error fetching engagement by event type:', error);
      throw error;
    }
  }

  /**
   * Get engagement within date range
   */
  async findByDateRange(startDate: Date, endDate: Date): Promise<EngagementData[]> {
    try {
      const result = await sql`
        SELECT id, user_id, simulation_id, event_type, event_data, timestamp, session_id
        FROM engagement_events
        WHERE timestamp >= ${startDate.toISOString()} AND timestamp <= ${endDate.toISOString()}
        ORDER BY timestamp DESC
      `;
      return result;
    } catch (error) {
      console.error('Error fetching engagement by date range:', error);
      throw error;
    }
  }

  /**
   * Get all engagement events
   */
  async findAll(limit?: number): Promise<EngagementData[]> {
    try {
      const query = limit
        ? sql`SELECT id, user_id, simulation_id, event_type, event_data, timestamp, session_id
             FROM engagement_events
             ORDER BY timestamp DESC
             LIMIT ${limit}`
        : sql`SELECT id, user_id, simulation_id, event_type, event_data, timestamp, session_id
             FROM engagement_events
             ORDER BY timestamp DESC`;

      return await query;
    } catch (error) {
      console.error('Error fetching all engagement:', error);
      throw error;
    }
  }

  /**
   * Create engagement event
   */
  async create(engagementData: CreateEngagementRequest): Promise<EngagementData> {
    try {
      const result = await sql`
        INSERT INTO engagement_events (
          user_id, simulation_id, event_type, event_data, session_id, timestamp
        )
        VALUES (
          ${engagementData.user_id},
          ${engagementData.simulation_id || null},
          ${engagementData.event_type},
          ${engagementData.event_data ? JSON.stringify(engagementData.event_data) : null},
          ${engagementData.session_id || null},
          NOW()
        )
        RETURNING id, user_id, simulation_id, event_type, event_data, timestamp, session_id
      `;
      return result[0];
    } catch (error) {
      console.error('Error creating engagement event:', error);
      throw error;
    }
  }

  /**
   * Bulk create engagement events
   */
  async bulkCreate(events: CreateEngagementRequest[]): Promise<EngagementData[]> {
    try {
      const values = events.map(e =>
        `('${e.user_id}', ${e.simulation_id ? `'${e.simulation_id}'` : 'NULL'}, '${e.event_type}', ${e.event_data ? `'${JSON.stringify(e.event_data)}'` : 'NULL'}, ${e.session_id ? `'${e.session_id}'` : 'NULL'}, NOW())`
      ).join(', ');

      const result = await sql`
        INSERT INTO engagement_events (user_id, simulation_id, event_type, event_data, session_id, timestamp)
        VALUES ${sql.raw(values)}
        RETURNING id, user_id, simulation_id, event_type, event_data, timestamp, session_id
      `;

      return result;
    } catch (error) {
      console.error('Error bulk creating engagement events:', error);
      throw error;
    }
  }

  /**
   * Delete engagement event
   */
  async delete(id: string): Promise<void> {
    try {
      await sql`
        DELETE FROM engagement_events
        WHERE id = ${id}
      `;
    } catch (error) {
      console.error('Error deleting engagement event:', error);
      throw error;
    }
  }

  /**
   * Delete engagement events by user
   */
  async deleteByUserId(userId: string): Promise<void> {
    try {
      await sql`
        DELETE FROM engagement_events
        WHERE user_id = ${userId}
      `;
    } catch (error) {
      console.error('Error deleting engagement events by user:', error);
      throw error;
    }
  }

  /**
   * Delete engagement events by simulation
   */
  async deleteBySimulationId(simulationId: string): Promise<void> {
    try {
      await sql`
        DELETE FROM engagement_events
        WHERE simulation_id = ${simulationId}
      `;
    } catch (error) {
      console.error('Error deleting engagement events by simulation:', error);
      throw error;
    }
  }

  /**
   * Get user engagement statistics
   */
  async getUserEngagementStats(userId: string): Promise<EngagementStats> {
    try {
      const totalResult = await sql`
        SELECT COUNT(*) as total, COUNT(DISTINCT session_id) as unique_sessions
        FROM engagement_events
        WHERE user_id = ${userId}
      `;

      const byTypeResult = await sql`
        SELECT event_type, COUNT(*) as count
        FROM engagement_events
        WHERE user_id = ${userId}
        GROUP BY event_type
      `;

      const eventsByType: Record<string, number> = {};
      byTypeResult.forEach((row: any) => {
        eventsByType[row.event_type] = parseInt(row.count);
      });

      const totalEvents = parseInt(totalResult[0].total);
      const uniqueSessions = parseInt(totalResult[0].unique_sessions) || 1;

      return {
        totalEvents,
        eventsByType,
        uniqueSessions,
        avgEventsPerSession: totalEvents / uniqueSessions,
      };
    } catch (error) {
      console.error('Error fetching user engagement stats:', error);
      throw error;
    }
  }

  /**
   * Get global engagement statistics
   */
  async getGlobalEngagementStats(startDate?: Date, endDate?: Date): Promise<EngagementStats> {
    try {
      let totalQuery, byTypeQuery;

      if (startDate && endDate) {
        totalQuery = sql`
          SELECT COUNT(*) as total, COUNT(DISTINCT session_id) as unique_sessions
          FROM engagement_events
          WHERE timestamp >= ${startDate.toISOString()} AND timestamp <= ${endDate.toISOString()}
        `;
        byTypeQuery = sql`
          SELECT event_type, COUNT(*) as count
          FROM engagement_events
          WHERE timestamp >= ${startDate.toISOString()} AND timestamp <= ${endDate.toISOString()}
          GROUP BY event_type
        `;
      } else {
        totalQuery = sql`
          SELECT COUNT(*) as total, COUNT(DISTINCT session_id) as unique_sessions
          FROM engagement_events
        `;
        byTypeQuery = sql`
          SELECT event_type, COUNT(*) as count
          FROM engagement_events
          GROUP BY event_type
        `;
      }

      const totalResult = await totalQuery;
      const byTypeResult = await byTypeQuery;

      const eventsByType: Record<string, number> = {};
      byTypeResult.forEach((row: any) => {
        eventsByType[row.event_type] = parseInt(row.count);
      });

      const totalEvents = parseInt(totalResult[0].total);
      const uniqueSessions = parseInt(totalResult[0].unique_sessions) || 1;

      return {
        totalEvents,
        eventsByType,
        uniqueSessions,
        avgEventsPerSession: totalEvents / uniqueSessions,
      };
    } catch (error) {
      console.error('Error fetching global engagement stats:', error);
      throw error;
    }
  }

  /**
   * Get daily engagement counts
   */
  async getDailyEngagementCounts(days: number = 30): Promise<Array<{date: string; count: number}>> {
    try {
      const result = await sql`
        SELECT DATE(timestamp) as date, COUNT(*) as count
        FROM engagement_events
        WHERE timestamp >= NOW() - INTERVAL '${days} days'
        GROUP BY DATE(timestamp)
        ORDER BY date DESC
      `;

      return result.map((row: any) => ({
        date: row.date,
        count: parseInt(row.count),
      }));
    } catch (error) {
      console.error('Error fetching daily engagement counts:', error);
      throw error;
    }
  }
}

// Export singleton instance
export const engagementRepository = new EngagementRepository();

File: backend-python/app/repositories/engagement_repository.py
"""
Engagement Repository
Database operations for engagement tracking entities
"""

import json
import logging
from typing import Any
from uuid import uuid4

from app.database import fetch, fetchone, execute, record_to_dict
from app.models.engagement import EngagementData, CreateEngagementRequest

logger = logging.getLogger(__name__)


class EngagementRepository:
    """Repository for engagement database operations."""

    async def find_by_id(self, engagement_id: str) -> EngagementData | None:
        """Find engagement event by ID."""
        try:
            query = """
                SELECT id, user_id, simulation_id, event_type, event_data, timestamp, created_at, updated_at
                FROM engagement_events WHERE id = $1
            """
            record = await fetchone(query, engagement_id)
            if not record:
                return None
            data = self._parse_json_fields(record_to_dict(record))
            return EngagementData(**data)
        except Exception as e:
            logger.error(f"Error finding engagement by ID: {e}")
            raise

    async def find_by_user_id(self, user_id: str) -> list[EngagementData]:
        """Find all engagement events for a user."""
        try:
            query = """
                SELECT id, user_id, simulation_id, event_type, event_data, timestamp, created_at, updated_at
                FROM engagement_events WHERE user_id = $1
                ORDER BY created_at DESC
            """
            records = await fetch(query, user_id)
            return [EngagementData(**self._parse_json_fields(dict(r))) for r in records]
        except Exception as e:
            logger.error(f"Error finding engagement by user ID: {e}")
            raise

    async def find_by_simulation_id(self, simulation_id: str) -> list[EngagementData]:
        """Find all engagement events for a simulation."""
        try:
            query = """
                SELECT id, user_id, simulation_id, event_type, event_data, timestamp, created_at, updated_at
                FROM engagement_events WHERE simulation_id = $1
                ORDER BY created_at ASC
            """
            records = await fetch(query, simulation_id)
            return [EngagementData(**self._parse_json_fields(dict(r))) for r in records]
        except Exception as e:
            logger.error(f"Error finding engagement by simulation ID: {e}")
            raise

    async def find_by_event_type(self, event_type: str) -> list[EngagementData]:
        """Find all engagement events of a type."""
        try:
            query = """
                SELECT id, user_id, simulation_id, event_type, event_data, timestamp, created_at, updated_at
                FROM engagement_events WHERE event_type = $1
                ORDER BY created_at DESC
            """
            records = await fetch(query, event_type)
            return [EngagementData(**self._parse_json_fields(dict(r))) for r in records]
        except Exception as e:
            logger.error(f"Error finding engagement by event type: {e}")
            raise

    async def create(self, engagement_data: CreateEngagementRequest | dict[str, Any]) -> EngagementData:
        """Create a new engagement event."""
        try:
            if isinstance(engagement_data, CreateEngagementRequest):
                data = engagement_data.model_dump()
            else:
                data = engagement_data

            engagement_id = str(uuid4())
            query = """
                INSERT INTO engagement_events (id, user_id, simulation_id, event_type, event_data, timestamp)
                VALUES ($1, $2, $3, $4, $5, NOW())
                RETURNING id, user_id, simulation_id, event_type, event_data, timestamp, created_at, updated_at
            """
            event_data = data.get("event_data")
            if event_data and not isinstance(event_data, str):
                event_data = json.dumps(event_data)

            record = await fetchone(
                query,
                engagement_id,
                data["user_id"],
                data.get("simulation_id"),
                data["event_type"],
                event_data,
            )
            result_data = self._parse_json_fields(record_to_dict(record))
            return EngagementData(**result_data)
        except Exception as e:
            logger.error(f"Error creating engagement event: {e}")
            raise

    async def delete_by_simulation_id(self, simulation_id: str) -> None:
        """Delete all engagement events for a simulation."""
        try:
            query = "DELETE FROM engagement_events WHERE simulation_id = $1"
            await execute(query, simulation_id)
        except Exception as e:
            logger.error(f"Error deleting engagement by simulation ID: {e}")
            raise

    async def delete_by_user_id(self, user_id: str) -> None:
        """Delete all engagement events for a user."""
        try:
            query = "DELETE FROM engagement_events WHERE user_id = $1"
            await execute(query, user_id)
        except Exception as e:
            logger.error(f"Error deleting engagement by user ID: {e}")
            raise

    async def get_user_metrics(self, user_id: str) -> dict[str, Any]:
        """Get engagement metrics for a user."""
        try:
            query = """
                SELECT
                    COUNT(*) as total_events,
                    COUNT(CASE WHEN event_type = 'simulation_start' THEN 1 END) as simulations_started,
                    COUNT(CASE WHEN event_type = 'simulation_complete' THEN 1 END) as simulations_completed,
                    COUNT(CASE WHEN event_type = 'message_sent' THEN 1 END) as messages_sent
                FROM engagement_events WHERE user_id = $1
            """
            record = await fetchone(query, user_id)
            if not record:
                return {
                    "totalEvents": 0,
                    "simulationsStarted": 0,
                    "simulationsCompleted": 0,
                    "messagesSent": 0,
                    "avgSessionDuration": 0,
                    "completionRate": 0,
                }
            data = dict(record)
            simulations_started = data.get("simulations_started", 0)
            simulations_completed = data.get("simulations_completed", 0)
            completion_rate = (simulations_completed / simulations_started * 100) if simulations_started > 0 else 0

            return {
                "totalEvents": data.get("total_events", 0),
                "simulationsStarted": simulations_started,
                "simulationsCompleted": simulations_completed,
                "messagesSent": data.get("messages_sent", 0),
                "avgSessionDuration": 0,  # Would need more complex calculation
                "completionRate": completion_rate,
            }
        except Exception as e:
            logger.error(f"Error getting user metrics: {e}")
            raise

    def _parse_json_fields(self, data: dict[str, Any]) -> dict[str, Any]:
        """Parse JSON string fields."""
        if data.get("event_data") and isinstance(data["event_data"], str):
            try:
                data["event_data"] = json.loads(data["event_data"])
            except json.JSONDecodeError:
                pass
        return data


# Singleton instance
engagement_repository = EngagementRepository()
